### PR TITLE
Move auto-indexing implementation out of Core API.

### PR DESCRIPTION
**Pull request by jakewins**
The purpose of this change is to make it possible to instantiate GraphDatabaseFacade
instances without also reconfiguring auto indexing(!). This, in turn, is part of work
to provide a GDS implementation to Procedures that does not allow `shutdown` 
and a few other low-level methods.

This is yet another step towards decoupling the legacy kernel code from
the cleaner Kernel API. In the past, auto-indexing was handled
(unsurprisingly) by the NodeManager god class. Auto indexing enabling
was done like:

```
Facade(enable tracking)
-> NodeManager.addTracker
-> (update list shared by reference to kernel)
```

Where the list of trackers was a generalized system for tracking properties,
which has never been used other than to add and remove this single auto
index tracker. The NodeManager had nothing to do with auto indexing, other
than that it held onto some state related to it. Auto indexes were then
updated whenever there was a change through this chain:

```
Facade(set property)
-> Kernel(trigger tracker)
-> LegacyPropertyTrackers
-> Facade(on tracker trigger)
-> DataWriteOps(idx update)
```

And finally, all this was configured as a side-effect of calling the GDF
constructor. This commit changes this, such that the Core API only acts
as an API, it delegates all auto index operations to a special AutoIndexing
service. The reason for doing it like so, rather than adding kernel
primitives is that AutoIndex configuration is set entirely at runtime, is
not transactional and is not propagated in clusters - it's simply runtime
state.

The AutoIndexing service is now shared between the Core API and the Kernel,
encapsulating the auto indexing logic. This means that enabling an auto
index now looks like:

```
Facade(enable tracking)
-> AutoIndexing#enabled
```

And property changes are propagated directly from the kernel without going
back up to the Core API:

```
Facade(set property)
-> Kernel
-> AutoIndexing#onChange
-> DataWriteOps(idx update)
```

### CODE REVIEW

**Comment by jakewins** [comment]
Marked as NOT_READY pending kernel team review

**Comment by tinwelint** [comment]
This PR is really, really sad, test-wise. I adore your commit messages btw

**Comment by jakewins** [comment]
@tinwelint green now :) If you guys are able to look at this on Monday, that'd be brilliant, me and @boggle are blocked on this, we've got another large branch that builds on this and we'd rather not keep piling on changes until we know this part is safe to rely on.

**Comment by tinwelint** [comment]
@jakewins I'll have a look asap

File: community/kernel/src/main/java/org/neo4j/kernel/PropertyTracker.java [hunk 1]
```diff
@@ -20,14 +20,13 @@
 package org.neo4j.kernel;
 
 import org.neo4j.graphdb.PropertyContainer;
+import org.neo4j.kernel.api.DataWriteOperations;
 
-/**
- * @deprecated This will be moved to internal packages in the next major release.
- */
-@Deprecated
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/PropertyTracker.java:27]: Deprecation removed and not moved? Is this interface even used?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/RelationshipAutoIndexerFacade.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import org.neo4j.kernel.impl.api.AutoIndexing;
+
+public class RelationshipAutoIndexerFacade extends AutoIndexerFacade<Relationship> implements RelationshipAutoIndexer
+{
+    private final Supplier<ReadableRelationshipIndex> idxSupplier;
+
+    public RelationshipAutoIndexerFacade( Supplier<ReadableRelationshipIndex> idxSupplier,
+                                          AutoIndexing.AutoIndexOperations autoIndexing )
+    {
+        super( idxSupplier::get, autoIndexing );
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/RelationshipAutoIndexerFacade.java:36]: nice trick

**Comment by tinwelint** [comment]
It's still little odd that auto indexing call stack is `DataWriteOperations` -> `EntityWriteOperations` -> `AutoIndexing` -> `DataWriteOperations` -> `LegacyIndexWriteOperations` so a bit back and forth. Although the changes in here makes things better than it was, so I won't let that hold it back.

**Comment by tinwelint** [comment]
@jakewins made some comments, but not super critical. Net change is good so merged it since you're blocked on it.

**Pull request merged**